File: app/src/test/java/com/example/recipeclipper/fake/FakeGroceryRepository.kt
package com.example.recipeclipper.fake

import com.example.recipeclipper.data.GroceryRepository
import com.example.recipeclipper.data.local.entity.GroceryItemEntity
import com.example.recipeclipper.data.model.Aisle
import com.example.recipeclipper.data.model.Aisles
import com.example.recipeclipper.data.model.GroceryItem
import com.example.recipeclipper.data.model.LanguageWords
import com.example.recipeclipper.data.model.NewGroceryLine
import com.example.recipeclipper.data.model.PlannedIngredients
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow

/**
 * A fake that models the list rather than only recording calls: [items] is real state, a new
 * line gets its aisle from the aisle table as the real repository does, and delete and restore
 * round-trip. [planned] is what [plannedIngredients] answers, filtered by day.
 */
class FakeGroceryRepository : GroceryRepository {

    val items = MutableStateFlow<List<GroceryItem>>(emptyList())
    var planned: List<PlannedIngredients> = emptyList()

    private var nextId = 1L

    override fun observeItems(): Flow<List<GroceryItem>> = items

    override suspend fun add(lines: List<NewGroceryLine>) {
        var order = (items.value.maxOfOrNull { it.sortOrder } ?: -1) + 1
        items.value = items.value + lines.filter { it.text.isNotBlank() }.map { line ->
            val text = line.text.trim()
            GroceryItem(
                id = nextId++,
                text = text,
                language = line.language,
                aisle = Aisles.of(text, LanguageWords.forTag(line.language)),
                checked = false,
                sortOrder = order++,
                recipeId = line.recipeId,
                plannedDay = line.plannedDay
            )
        }
    }

    override suspend fun setChecked(ids: List<Long>, checked: Boolean) {
        items.value = items.value.map { if (it.id in ids) it.copy(checked = checked) else it }
    }

    override suspend fun setAisle(ids: List<Long>, aisle: Aisle) {
        items.value = items.value.map { if (it.id in ids) it.copy(aisle = aisle) else it }
    }

    override suspend fun delete(ids: List<Long>): GroceryRepository.DeletedItems? =
        remove { it.id in ids }

    override suspend fun clearChecked(): GroceryRepository.DeletedItems? = remove { it.checked }

    private fun remove(which: (GroceryItem) -> Boolean): GroceryRepository.DeletedItems? {
        val gone = items.value.filter(which).ifEmpty { return null }
        items.value = items.value.filterNot(which)
        return GroceryRepository.DeletedItems(gone.map { it.toEntity() })
    }

    override suspend fun restore(deleted: GroceryRepository.DeletedItems) {
        items.value = (items.value + deleted.entities.map { it.toDomain() }).sortedBy { it.sortOrder }
    }

    override suspend fun plannedIngredients(start: Long, end: Long): List<PlannedIngredients> =
        planned.filter { it.day in start..end }

    private fun GroceryItem.toEntity() = GroceryItemEntity(
        id = id, text = text, language = language, aisle = aisle.key, checked = checked,
        sortOrder = sortOrder, recipeId = recipeId, plannedDay = plannedDay, updatedAt = 0
    )

    private fun GroceryItemEntity.toDomain() = GroceryItem(
        id = id, text = text, language = language, aisle = Aisle.fromKey(aisle), checked = checked,
        sortOrder = sortOrder, recipeId = recipeId, plannedDay = plannedDay
    )
}

File: ios/RecipeClipperTests/Fakes/FakeGroceryRepository.swift
import Combine
import Foundation
@testable import RecipeClipper

/// A fake that models the list rather than only recording calls (Android's
/// FakeGroceryRepository): `items` is real state, a new line gets its aisle from the aisle
/// table as the real repository does, and delete and restore round-trip. `planned` is what
/// `plannedIngredients` answers, filtered by day.
final class FakeGroceryRepository: GroceryRepository {
    let items = CurrentValueSubject<[GroceryItem], Never>([])
    var planned: [PlannedIngredients] = []

    private var nextId: Int64 = 1

    func observeItems() -> AnyPublisher<[GroceryItem], Never> { items.eraseToAnyPublisher() }

    func add(_ lines: [NewGroceryLine]) async {
        var order = (items.value.map(\.sortOrder).max() ?? -1) + 1
        var added: [GroceryItem] = []
        for line in lines where !line.text.kIsBlank {
            let text = line.text.kTrimmed
            added.append(GroceryItem(
                id: nextId, text: text, language: line.language,
                aisle: Aisles.of(text, words: LanguageWords.forTag(line.language)), checked: false,
                sortOrder: order, recipeId: line.recipeId, plannedDay: line.plannedDay
            ))
            nextId += 1
            order += 1
        }
        items.value += added
    }

    func setChecked(_ ids: [Int64], checked: Bool) async {
        items.value = items.value.map { item in
            var i = item
            if ids.contains(i.id) { i.checked = checked }
            return i
        }
    }

    func setAisle(_ ids: [Int64], aisle: Aisle) async {
        items.value = items.value.map { item in
            var i = item
            if ids.contains(i.id) { i.aisle = aisle }
            return i
        }
    }

    func delete(_ ids: [Int64]) async -> DeletedGroceries? { remove { ids.contains($0.id) } }

    func clearChecked() async -> DeletedGroceries? { remove(\.checked) }

    private func remove(_ which: (GroceryItem) -> Bool) -> DeletedGroceries? {
        let gone = items.value.filter(which)
        if gone.isEmpty { return nil }
        items.value.removeAll(where: which)
        return DeletedGroceries(items: gone.map {
            GroceryItemRecord(
                id: $0.id, text: $0.text, language: $0.language, aisle: $0.aisle.key, checked: $0.checked,
                sortOrder: $0.sortOrder, recipeId: $0.recipeId, plannedDay: $0.plannedDay, updatedAt: 0
            )
        })
    }

    func restore(_ deleted: DeletedGroceries) async {
        items.value = (items.value + deleted.items.map(\.domain)).sorted { $0.sortOrder < $1.sortOrder }
    }

    func plannedIngredients(start: Int64, end: Int64) async -> [PlannedIngredients] {
        planned.filter { $0.day >= start && $0.day <= end }
    }
}
